applications/FluidDynamicsApplication/custom_utilities/statistics_data.h

```diff
@@ -0,0 +1,328 @@
+     *  @param NumberOfMeasurements How many times this function has been called so far (including this one).
+     */
+    void UpdateMeasurement(
+        const Element& rElement,
+        const PointerVector<StatisticsSampler>& rStatisticsSamplers,
+        const PointerVector<StatisticsSampler>& rHigherOrderStatistics,
+        ValueContainerType& rUpdate,
+        std::size_t NumMeasurements)
+    {
+        KRATOS_DEBUG_ERROR_IF(NumMeasurements == 0)
+        << "Trying to update statistics, but providied number of recorded steps is zero" << std::endl;
+
+        const Geometry<Node<3>> &r_geometry = rElement.GetGeometry();
+        const GeometryData::IntegrationMethod integration_method = rElement.GetIntegrationMethod();
+        Matrix shape_functions;
+        typename Geometry<Node<3>>::ShapeFunctionsGradientsType shape_gradients;
+        this->CalculateGeometryData(r_geometry, integration_method, shape_functions, shape_gradients);
+
+        for (unsigned int g = 0; g < shape_functions.size1(); g++)
+        {
+            auto N = row(shape_functions, g);
+            auto &rDN_DN = shape_gradients[g];
+
+            auto it_update_buffer = rUpdate.begin();
+            for (auto it_sampler = rStatisticsSamplers.begin(); it_sampler != rStatisticsSamplers.end(); ++it_sampler)
+            {
+                it_sampler->SampleDataPoint(r_geometry, N, rDN_DN, it_update_buffer);
+            }
+
+            if (NumMeasurements > 1) { // Second order and higher statistics start from the second iteration
+                // loop on higer order statistics. They require the already written rMeasurements, (const) mData and the number of steps as input
+                for (auto it_statistic = rHigherOrderStatistics.begin(); it_statistic != rHigherOrderStatistics.end(); ++it_statistic)
+                {
+                    it_statistic->SampleDataPoint(it_update_buffer, IntegrationPointData(g), rUpdate, NumMeasurements);
+                }
+            }
+
+            for (unsigned int i = 0; i < rUpdate.size(); i++)
+            {
+                mData(g,i) += rUpdate[i];
+            }
+        }
+    }
+
+    /// Write computed statistics to output file.
+    /** @param rOutputStream output file stream.
+     *  @param rElement Element instance this container stores data for.
+     *  @param rRecordedStatistics List of first order statistics stored in this container.
+     *  @param rHigherOrderStatistics List of higher order statistics stored in this container.
+     *  @param NumberOfMeasurements How many realizations have been recorded up to this point.
+     *  @param rSeparator Separator to use in the output file.
+     */
+    void WriteToCSVOutput(
+        std::ofstream& rOutputStream,
+        const Element& rElement,
+        const PointerVector<StatisticsSampler>& rRecordedStatistics,
+        const PointerVector<StatisticsSampler>& rHigherOrderStatistics,
+        std::size_t NumberOfMeasurements,
```

Comment:
(very) minor, but you might want to have this const

Reply:
I didn't bother with it because it is a local variable (not a reference like the others), but it makes sense.